### ISSUE TITLE
Yaml files for FFA-LoRA, FedDPA-LoRA

### ISSUE THREAD

**Issue by EnsiehKhazaei** [Pengxin-Guo/FedSA-LoRA#12]
Hi @Pengxin-Guo, 
Congratulations on your great work. I am wondering ,could you provide the yaml files for FFA-LoRA and FedDPA-LoRA methods, or let us know how to change the yaml file to reproduce your results for these methods  
Thank you

**Comment by Pengxin-Guo** [comment]
Hi,
For FFA-LoRA, you can set `freeze_A` to `True` to reproduce the results.
For FedDPA-LoRA, you need to add a few lines of code in https://github.com/Pengxin-Guo/FedSA-LoRA/blob/5e616a06e806a8804e09820ddebfcbd2363f0a2f/federatedscope/glue/model/model_builder.py#L36 as follows:

```
if config.federate.dual_lora == True:
        from peft import LoraConfig, TaskType, VeraConfig
        kwargs = args
        kwargs.pop('adapter_package', None)
        adapter_method = kwargs.pop('adapter_method', 'lora')
        if adapter_method == 'vera':
            peft_config = VeraConfig(task_type=TaskType.SEQ_CLS, **kwargs)
        else:
            peft_config = LoraConfig(task_type=TaskType.SEQ_CLS, **kwargs)
        model.model.add_adapter("local", peft_config)
        model.model.base_model.set_adapter(["default", "local"])
```
Then, make sure to adjust the parameters for the `dual_lora` and `personalization` accordingly.

If you have any further questions, feel free to ask!

**Comment by PabloE-KAIST** [comment]
@Pengxin-Guo, could you kindly provide an example of setting the personalization config suitably to reproduce FedDPA?

**Comment by EnsiehKhazaei** [comment]
> Hi, For FFA-LoRA, you can set `freeze_A` to `True` to reproduce the results. For FedDPA-LoRA, you need to add a few lines of code in
> 
> [FedSA-LoRA/federatedscope/glue/model/model_builder.py](https://github.com/Pengxin-Guo/FedSA-LoRA/blob/5e616a06e806a8804e09820ddebfcbd2363f0a2f/federatedscope/glue/model/model_builder.py#L36)
> 
> Line 36 in [5e616a0](/Pengxin-Guo/FedSA-LoRA/commit/5e616a06e806a8804e09820ddebfcbd2363f0a2f)
> 
>   
> as follows:
> ```
> if config.federate.dual_lora == True:
>         from peft import LoraConfig, TaskType, VeraConfig
>         kwargs = args
>         kwargs.pop('adapter_package', None)
>         adapter_method = kwargs.pop('adapter_method', 'lora')
>         if adapter_method == 'vera':
>             peft_config = VeraConfig(task_type=TaskType.SEQ_CLS, **kwargs)
>         else:
>             peft_config = LoraConfig(task_type=TaskType.SEQ_CLS, **kwargs)
>         model.model.add_adapter("local", peft_config)
>         model.model.base_model.set_adapter(["default", "local"])
> ```
> 
> Then, make sure to adjust the parameters for the `dual_lora` and `personalization` accordingly.
> 
> If you have any further questions, feel free to ask!

Thank you so much, I also have a few questions regarding the NLG task:
1. Do you use for all methods for dataset GSM8K, a learning rate of 0.005 and for dataset CodesearchNet, a learning rate of 0.01 for all methods? If different LoRA and FFA use other learning rate, could you provide learning rate for them
2. For how many seeds, do you run NLG tasks?
3. Could you also share your YAML file for CodeSearchNet to reproduce the results

**Comment by Pengxin-Guo** [comment]
> [@Pengxin-Guo](https://github.com/Pengxin-Guo), could you kindly provide an example of setting the personalization config suitably to reproduce FedDPA?

Sure, below is the `personalization` config for FedDPA:
```
personalization: 
  local_param: ['local']    # for FedDPA
```

**Comment by Pengxin-Guo** [comment]
> > Hi, For FFA-LoRA, you can set `freeze_A` to `True` to reproduce the results. For FedDPA-LoRA, you need to add a few lines of code in
> > [FedSA-LoRA/federatedscope/glue/model/model_builder.py](https://github.com/Pengxin-Guo/FedSA-LoRA/blob/5e616a06e806a8804e09820ddebfcbd2363f0a2f/federatedscope/glue/model/model_builder.py#L36)
> > Line 36 in [5e616a0](/Pengxin-Guo/FedSA-LoRA/commit/5e616a06e806a8804e09820ddebfcbd2363f0a2f)
> > as follows:
> > ```
> > if config.federate.dual_lora == True:
> >         from peft import LoraConfig, TaskType, VeraConfig
> >         kwargs = args
> >         kwargs.pop('adapter_package', None)
> >         adapter_method = kwargs.pop('adapter_method', 'lora')
> >         if adapter_method == 'vera':
> >             peft_config = VeraConfig(task_type=TaskType.SEQ_CLS, **kwargs)
> >         else:
> >             peft_config = LoraConfig(task_type=TaskType.SEQ_CLS, **kwargs)
> >         model.model.add_adapter("local", peft_config)
> >         model.model.base_model.set_adapter(["default", "local"])
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then, make sure to adjust the parameters for the `dual_lora` and `personalization` accordingly.
> > If you have any further questions, feel free to ask!
> 
> Thank you so much, I also have a few questions regarding the NLG task:
> 
> 1. Do you use for all methods for dataset GSM8K, a learning rate of 0.005 and for dataset CodesearchNet, a learning rate of 0.01 for all methods? If different LoRA and FFA use other learning rate, could you provide learning rate for them
> 2. For how many seeds, do you run NLG tasks?
> 3. Could you also share your YAML file for CodeSearchNet to reproduce the results

1. Yes, all methods use the same learning rate of 0.005 for GSM8K and 0.001 for CodesearchNet.
2. Only one seed: 0.
3. Below is the YAML file for CodeSearchNet:
```
use_gpu: True
device: 0
eval_device: 0
early_stop:
  patience: 0
federate:
  # freeze_A: True   # for FFA-LoRA
  mode: standalone
  client_num: 6
  total_round_num: 100
  save_to: "/path/llama_csn_fed_30*500.ckpt"
  save_freq: 100
  share_local_model: True
  online_aggr: False
personalization: 
  local_param: ['lora_B']    # for ours
data:
  root: /path/csn_data/
  type: 'code_search_net@llm'
  splitter: 'meta'
  subsample: 0.05
llm:
  cache: 
      model: /path/model_cache/
  tok_len: 650
  chat:
    max_len: 1000
  adapter:
    use: True
    args: [ { 'adapter_package': 'peft', 'adapter_method': 'lora', 'r': 8, 'lora_alpha': 16, 'lora_dropout': 0.05 } ]
dataloader:
  batch_size: 1
model:
  type: 'meta-llama/Meta-Llama-3-8B@huggingface_llm'
train:
  local_update_steps: 30
  batch_or_epoch: batch
  optimizer:
    lr: 0.001
    weight_decay: 0.0
  is_enable_half: True
criterion:
  type: CrossEntropyLoss
trainer:
  type: llmtrainer
eval:
  freq: 40
  metrics: ['acc']
  best_res_update_round_wise_key: val_acc
  count_flops: False
```

**Comment by EnsiehKhazaei** [comment]
> > > Hi, For FFA-LoRA, you can set `freeze_A` to `True` to reproduce the results. For FedDPA-LoRA, you need to add a few lines of code in
> > > [FedSA-LoRA/federatedscope/glue/model/model_builder.py](https://github.com/Pengxin-Guo/FedSA-LoRA/blob/5e616a06e806a8804e09820ddebfcbd2363f0a2f/federatedscope/glue/model/model_builder.py#L36)
> > > Line 36 in [5e616a0](/Pengxin-Guo/FedSA-LoRA/commit/5e616a06e806a8804e09820ddebfcbd2363f0a2f)
> > > as follows:
> > > ```
> > > if config.federate.dual_lora == True:
> > >         from peft import LoraConfig, TaskType, VeraConfig
> > >         kwargs = args
> > >         kwargs.pop('adapter_package', None)
> > >         adapter_method = kwargs.pop('adapter_method', 'lora')
> > >         if adapter_method == 'vera':
> > >             peft_config = VeraConfig(task_type=TaskType.SEQ_CLS, **kwargs)
> > >         else:
> > >             peft_config = LoraConfig(task_type=TaskType.SEQ_CLS, **kwargs)
> > >         model.model.add_adapter("local", peft_config)
> > >         model.model.base_model.set_adapter(["default", "local"])
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Then, make sure to adjust the parameters for the `dual_lora` and `personalization` accordingly.
> > > If you have any further questions, feel free to ask!
> > 
> > 
> > Thank you so much, I also have a few questions regarding the NLG task:
> > 
> > 1. Do you use for all methods for dataset GSM8K, a learning rate of 0.005 and for dataset CodesearchNet, a learning rate of 0.01 for all methods? If different LoRA and FFA use other learning rate, could you provide learning rate for them
> > 2. For how many seeds, do you run NLG tasks?
> > 3. Could you also share your YAML file for CodeSearchNet to reproduce the results
> 
> 1. Yes, all methods use the same learning rate of 0.005 for GSM8K and 0.001 for CodesearchNet.
> 2. Only one seed: 0.
> 3. Below is the YAML file for CodeSearchNet:
> 
> ```
> use_gpu: True
> device: 0
> eval_device: 0
> early_stop:
>   patience: 0
> federate:
>   # freeze_A: True   # for FFA-LoRA
>   mode: standalone
>   client_num: 6
>   total_round_num: 100
>   save_to: "/path/llama_csn_fed_30*500.ckpt"
>   save_freq: 100
>   share_local_model: True
>   online_aggr: False
> personalization: 
>   local_param: ['lora_B']    # for ours
> data:
>   root: /path/csn_data/
>   type: 'code_search_net@llm'
>   splitter: 'meta'
>   subsample: 0.05
> llm:
>   cache: 
>       model: /path/model_cache/
>   tok_len: 650
>   chat:
>     max_len: 1000
>   adapter:
>     use: True
>     args: [ { 'adapter_package': 'peft', 'adapter_method': 'lora', 'r': 8, 'lora_alpha': 16, 'lora_dropout': 0.05 } ]
> dataloader:
>   batch_size: 1
> model:
>   type: 'meta-llama/Meta-Llama-3-8B@huggingface_llm'
> train:
>   local_update_steps: 30
>   batch_or_epoch: batch
>   optimizer:
>     lr: 0.001
>     weight_decay: 0.0
>   is_enable_half: True
> criterion:
>   type: CrossEntropyLoss
> trainer:
>   type: llmtrainer
> eval:
>   freq: 40
>   metrics: ['acc']
>   best_res_update_round_wise_key: val_acc
>   count_flops: False
> ```

Thank you so much for sharing the codesearchnet YAML file. I use this YAML file, and I ran "python federatedscope/main.py --cfg federatedscope/llm/yamls/fedsa-lora-codesearchnet.yaml", but I got stuck in the evaluation part (I got some error, and it does not proceed anymore). Could you help me with how to do an evaluation for this dataset every 40 rounds?

**Comment by Xiaoman916** [comment]
有对比实验吗？

**Comment by minhdn2k2** [comment]
Hi @Pengxin-Guo ,

This is an excellent work on FedSA-LoRA. I am currently reproducing the code, and I have a question regarding the evaluation protocol.

My current understanding is that training data are partitioned across clients, while validation set is kept at the server side and used for evaluation after each communication round. However, I'm still not fully sure how FedSA-LoRA is actually evaluated, since it includes personalized components such as local LoRA-B and classifier.

Because my understanding of this area is still limited, could you let me know how the evaluation is performed in your implementation on FedSA-LORA.
